Dockerize the engineering blog

Crunch09 opened the pull request:
As discussed we want to dockerize the engineering blog to make it easier for non-ruby people in the tech org to contribute.
This PR adds a `Dockerfile` to achieve that goal. Run it with:
```bash
$ docker build --no-cache -t deliveroo-engineering .
$ docker run --rm --name deliveroo-engineering -p 4000:4000 -v $(pwd):/usr/src/app/deliveroo.engineering -it deliveroo-engineering bundle exec jekyll serve -H 0.0.0.0
```
Access the blog now on `http://localhost:4000`.
Local file changes (updating posts, creating posts) will already trigger a jekyll rebuild within the running container.

- [x] Move commands to a Script / Makefile
- [x] Update documentation
- [ ] Test with non-ruby colleagues

In a follow-up PR i'll then add support for [jekyll-admin](https://github.com/jekyll/jekyll-admin).

@jphastings What do you think?

_Fixes [EB-14](https://deliveroo.atlassian.net/browse/EB-14)_

kapowaz commented:
👏 This is a great idea! Whilst a Jekyll blog might not exist forever, it’s definitely worth making it easier for non-eng folk to set up and run locally.

I’d suggest using [this Makefile](https://github.com/deliveroo/restaurant-hub-app/blob/staging/Makefile) as a template for making these commands easier to run as probably the pattern we’ll follow going forward.

jphastings commented on Dockerfile:
```diff
@@ -0,0 +1,9 @@
+RUN apk add --no-cache build-base
+RUN gem install bundler
+
+COPY . /usr/src/app/deliveroo.engineering
```
A nice trick with docker is to copy _just_ the stuff you'll need for doing a dependency install first, then copying everything else, eg.

```docker
COPY Gemfile Gemfile.lock /usr/src/app/deliveroo.engineering
WORKDIR /usr/src/app/deliveroo.engineering
RUN gem install bundler && bundle install
COPY . /usr/src/app/deliveroo.engineering
```

This means that if files that _aren't_ the `Gemfile[.lock]` are edited, only the last step of the Dockerfile needs to be repeated (docker can operate from the cache for the `bundle install`), which makes it much faster when doing lots of edits 😊

jphastings commented on Dockerfile:
```diff
@@ -0,0 +1,9 @@
+COPY . /usr/src/app/deliveroo.engineering
+WORKDIR /usr/src/app/deliveroo.engineering
+
+RUN bundle install
```
I think you might want your last line to be running jekyll:

```docker
ENTRYPOINT ["jekyll", "start"]
```

Crunch09 commented:
Thanks @jphastings and @kapowaz I updated the Dockerfile, added a Makefile and added documentation.
Running it is just basically two commands:
```sh
$ make build
$ make run
```

jphastings commented on Makefile:
```diff
@@ -0,0 +1,8 @@
+	  @echo Building deliveroo-engineering
+	  docker build -t deliveroo-engineering .
+
+run:
```
You can make this line `run: build`, which will execute the build before every run — docker is smart so won't build again if it doesn't need to (but will if the files added to the repo have been edited). In combination with my suggestion above it should be a slowish start first time around/when you make a change to the `Gemfile`, then rapid after that!

Crunch09 replied:
Thank you, I'll do that!

kapowaz replied:
Is that definitely the case @jphastings? I know from my experience running other projects of ours which use Docker that running `make build run` is always slower than just `make run`, regardless of whether you’ve already previously run `build`.

bencord0 replied:
Docker isn't that smart about build caching, if you have a `RUN` after a `COPY`, then you don't get any of the benefits.

It might also be worth making use of `.PHONY` targets, e.g. `.PHONY: build run` to prevent silly things happening if a file named `build` or `run` exists. This, combined with `run: build` will ensure that the `docker build` command is run before `docker run` on a simple `make run`.

Now, do we have to do that?
If we remove line 9 in the Dockerfile and only require docker to rebuild if the Gemfile/Gemfile.lock change, then decoupling `docker build` and `docker run` steps make sense.

Always use docker run, unless you have updated the gems.

The only downside is that the resulting Dockerfile is not suitable for production. This is fine since we currently use gh-pages to host this repository.

Something something #HopperFTW! but this is a public comment/repository, and we don't need to require that all contributors have access to the clusters.

jphastings replied:
Hey @bencord0, this seems great advice! Would you be able to work with @Crunch09 to implement these changes? A lot of this is out of my wheelhouse and I'd spend time fumbling in the dark 😅

bencord0 replied:
This patch should do it.

I've decided not to make `run` depend on `build`, I'm happy that the current instructions in the `README` are enough to get the local environment working.

```
diff --git a/Dockerfile b/Dockerfile
index 748b666..6abd6a7 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -6,7 +6,6 @@ COPY Gemfile Gemfile.lock /usr/src/app/deliveroo.engineering/
 WORKDIR /usr/src/app/deliveroo.engineering
 RUN gem install bundler && bundle install -j8

-COPY . /usr/src/app/deliveroo.engineering
 EXPOSE 4000
 ENTRYPOINT ["jekyll"]
 CMD ["serve", "-w", "-t", "-H", "0.0.0.0"]
diff --git a/Makefile b/Makefile
index 1f252dc..57f41bd 100644
--- a/Makefile
+++ b/Makefile
@@ -6,3 +6,5 @@ build:
 run:
          @echo Starting jekyll server on port 4000
          docker run -p 4000:4000 -v $(shell pwd):/usr/src/app/deliveroo.engineering deliveroo-engineering
+
+.PHONY: build run
```

Crunch09 replied:
Thank you! 🙇  @bencord0 Will try to update the instructions today and then we can finally get it merged

jphastings commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+WORKDIR /usr/src/app/deliveroo.engineering
+RUN gem install bundler && bundle install -j8
+
+COPY . /usr/src/app/deliveroo.engineering
```
Do you need this line? Given your `make run` below maps the local folder on top of this one you could use the [`VOLUME`](https://docs.docker.com/engine/reference/builder/#volume) command instead so the docker file doesn't need to rebuild every time it starts (only if you change the `Gemfile` or `Gemfile.lock`)

Crunch09 replied:
Good point, thanks! Didn't really know how `VOLUME` works but i'll look into it and amend this PR.
